// @sphereon/pex@2.0.1/dist/browser/lib/evaluation/evaluationClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EvaluationClient = void 0;
const ssi_types_1 = require("@sphereon/ssi-types");
const ConstraintUtils_1 = require("../ConstraintUtils");
const Messages_1 = __importDefault(require("../types/Messages"));
const utils_1 = require("../utils");
const handlers_1 = require("./handlers");
const DEFAULT_LIMIT_DISCLOSURE_TYPES = [ssi_types_1.IProofType.BbsBlsSignatureProof2020];
class EvaluationClient {
    constructor() {
        this.failed_catched = {
            tag: 'root',
            status: ConstraintUtils_1.Status.ERROR,
            message: Messages_1.default.UNKNOWN_EXCEPTION,
            stacktrace: '',
        };
        this._results = [];
        this._wrappedVcs = [];
        this._presentationSubmission = {};
        this._dids = [];
        this._limitDisclosureSignatureSuites = DEFAULT_LIMIT_DISCLOSURE_TYPES;
        this._restrictToDIDMethods = [];
        this._generatePresentationSubmission = true;
    }
    evaluate(pd, wvcs, opts) {
        this._restrictToDIDMethods = (opts === null || opts === void 0 ? void 0 : opts.restrictToDIDMethods) ? (0, utils_1.uniformDIDMethods)(opts === null || opts === void 0 ? void 0 : opts.restrictToDIDMethods) : [];
        this._dids = (opts === null || opts === void 0 ? void 0 : opts.holderDIDs) ? (0, utils_1.filterToRestrictedDIDs)(opts.holderDIDs, this._restrictToDIDMethods) : [];
        this._limitDisclosureSignatureSuites = opts === null || opts === void 0 ? void 0 : opts.limitDisclosureSignatureSuites;
        this._restrictToFormats = opts === null || opts === void 0 ? void 0 : opts.restrictToFormats;
        this._generatePresentationSubmission = (opts === null || opts === void 0 ? void 0 : opts.generatePresentationSubmission) !== undefined ? opts.generatePresentationSubmission : true;
        if (opts === null || opts === void 0 ? void 0 : opts.presentationSubmission) {
            this._presentationSubmission = opts.presentationSubmission;
            // this._requirePresentationSubmission = true;
        }
        let currentHandler = this.initEvaluationHandlers();
        currentHandler === null || currentHandler === void 0 ? void 0 : currentHandler.handle(pd, wvcs);
        while (currentHandler === null || currentHandler === void 0 ? void 0 : currentHandler.hasNext()) {
            currentHandler = currentHandler.getNext();
            try {
                currentHandler === null || currentHandler === void 0 ? void 0 : currentHandler.handle(pd, wvcs);
            }
            catch (e) {
                this.failed_catched.message += e.message;
                this.failed_catched.stacktrace = e;
                throw this.failed_catched;
            }
        }
    }
    get results() {
        return this._results;
    }
    get dids() {
        return this._dids;
    }
    set dids(dids) {
        this._dids = dids;
    }
    assertPresentationSubmission() {
        if (!this.generatePresentationSubmission && (!this.presentationSubmission || Object.keys(this.presentationSubmission).length === 0)) {
            throw Error('No presentation submission present, but required option was set');
        }
    }
    get generatePresentationSubmission() {
        return this._generatePresentationSubmission;
    }
    set generatePresentationSubmission(value) {
        this._generatePresentationSubmission = value;
    }
    get presentationSubmission() {
        return this._presentationSubmission;
    }
    set presentationSubmission(presentationSubmission) {
        this._presentationSubmission = presentationSubmission;
    }
    get wrappedVcs() {
        return this._wrappedVcs;
    }
    set wrappedVcs(wrappedVcs) {
        this._wrappedVcs = wrappedVcs;
    }
    get limitDisclosureSignatureSuites() {
        return this._limitDisclosureSignatureSuites || DEFAULT_LIMIT_DISCLOSURE_TYPES;
    }
    set limitDisclosureSignatureSuites(limitDisclosureSignatureSuites) {
        this._limitDisclosureSignatureSuites = limitDisclosureSignatureSuites;
    }
    get restrictToDIDMethods() {
        return this._restrictToDIDMethods;
    }
    set restrictToDIDMethods(value) {
        this._restrictToDIDMethods = (0, utils_1.uniformDIDMethods)(value);
    }
    hasRestrictToDIDMethods() {
        return this.restrictToDIDMethods && this.restrictToDIDMethods.length > 0;
    }
    get restrictToFormats() {
        return this._restrictToFormats;
    }
    set restrictToFormats(value) {
        this._restrictToFormats = value;
    }
    initEvaluationHandlers() {
        const uriEvaluation = new handlers_1.UriEvaluationHandler(this);
        uriEvaluation
            .setNext(new handlers_1.DIDRestrictionEvaluationHandler(this))
            .setNext(new handlers_1.FormatRestrictionEvaluationHandler(this))
            .setNext(new handlers_1.InputDescriptorFilterEvaluationHandler(this))
            .setNext(new handlers_1.PredicateRelatedFieldEvaluationHandler(this))
            .setNext(new handlers_1.LimitDisclosureEvaluationHandler(this))
            .setNext(new handlers_1.SubjectIsIssuerEvaluationHandler(this))
            .setNext(new handlers_1.SubjectIsHolderEvaluationHandler(this))
            .setNext(new handlers_1.SameSubjectEvaluationHandler(this))
            .setNext(new handlers_1.MarkForSubmissionEvaluationHandler(this));
        return uriEvaluation;
    }
}
exports.EvaluationClient = EvaluationClient;
