// @sphereon/pex@2.0.1/dist/main/lib/types/SSITypesBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSITypesBuilder = void 0;
const ssi_types_1 = require("@sphereon/ssi-types");
const utils_1 = require("../utils");
const Internal_types_1 = require("./Internal.types");
class SSITypesBuilder {
    static modelEntityToInternalPresentationDefinitionV1(p) {
        const pd = SSITypesBuilder.createCopyAndModifyPresentationDefinition(p);
        return new Internal_types_1.InternalPresentationDefinitionV1(pd.id, pd.input_descriptors, pd.format, pd.name, pd.purpose, pd.submission_requirements);
    }
    static modelEntityInternalPresentationDefinitionV2(p) {
        const pd = SSITypesBuilder.createCopyAndModifyPresentationDefinition(p);
        return new Internal_types_1.InternalPresentationDefinitionV2(pd.id, pd.input_descriptors, pd.format, pd.frame, pd.name, pd.purpose, pd.submission_requirements);
    }
    static createCopyAndModifyPresentationDefinition(p) {
        const pd = JSON.parse(JSON.stringify(p));
        utils_1.JsonPathUtils.changePropertyNameRecursively(pd, '_const', 'const');
        utils_1.JsonPathUtils.changePropertyNameRecursively(pd, '_enum', 'enum');
        utils_1.JsonPathUtils.changeSpecialPathsRecursively(pd);
        return pd;
    }
    static mapExternalVerifiablePresentationToWrappedVP(presentation) {
        return ssi_types_1.CredentialMapper.toWrappedVerifiablePresentation(presentation);
    }
    static mapExternalVerifiableCredentialsToWrappedVcs(verifiableCredentials) {
        return ssi_types_1.CredentialMapper.toWrappedVerifiableCredentials(Array.isArray(verifiableCredentials) ? verifiableCredentials : [verifiableCredentials]);
    }
    static toInternalPresentationDefinition(presentationDefinition) {
        const presentationDefinitionCopy = JSON.parse(JSON.stringify(presentationDefinition));
        const versionResult = (0, utils_1.definitionVersionDiscovery)(presentationDefinitionCopy);
        if (versionResult.error)
            throw versionResult.error;
        if (versionResult.version == Internal_types_1.PEVersion.v1) {
            return SSITypesBuilder.modelEntityToInternalPresentationDefinitionV1(presentationDefinitionCopy);
        }
        return SSITypesBuilder.modelEntityInternalPresentationDefinitionV2(presentationDefinitionCopy);
    }
}
exports.SSITypesBuilder = SSITypesBuilder;
