// @sphereon/pex@2.0.1/dist/browser/lib/validation/bundlers/frameVB.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameVB = void 0;
const utils_1 = require("../../utils");
const validationBundler_1 = require("./validationBundler");
class FrameVB extends validationBundler_1.ValidationBundler {
    constructor(parentTag) {
        super(parentTag, 'frame');
        this.frameIsValidMsg = 'frame value is not valid';
    }
    getValidations(frame) {
        let validations = [];
        validations = [...validations, ...this.getMyValidations(frame)];
        return validations;
    }
    getMyValidations(frame) {
        return [
            {
                tag: this.getMyTag(),
                target: frame,
                predicate: FrameVB.frameIsValid,
                message: this.frameIsValidMsg,
            },
        ];
    }
    getMyTag() {
        return this.parentTag + '.' + this.myTag;
    }
    /**
     * this is based on https://github.com/digitalbazaar/jsonld.js/blob/main/lib/frame.js
     * @param frame
     */
    static frameIsValid(frame) {
        if (!frame || Array.isArray(frame) || !(typeof frame === 'object')) {
            return false;
        }
        const fr = frame;
        if (fr && fr['@id']) {
            for (const id of utils_1.ObjectUtils.asArray(frame['@id'])) {
                // @id must be wildcard or an IRI
                if (!(utils_1.ObjectUtils.isObject(id) || utils_1.ObjectUtils.isUrlAbsolute(id)) || (utils_1.ObjectUtils.isString(id) && id.indexOf('_:') === 0)) {
                    return false;
                }
            }
        }
        //
        if (fr['@types']) {
            for (const type of utils_1.ObjectUtils.asArray(frame['@types'])) {
                // @id must be wildcard or an IRI
                if (!(utils_1.ObjectUtils.isObject(type) || utils_1.ObjectUtils.isUrlAbsolute(type)) || (utils_1.ObjectUtils.isString(type) && type.indexOf('_:') === 0)) {
                    return false;
                }
            }
        }
        return true;
    }
}
exports.FrameVB = FrameVB;
