// @sphereon/pex@2.0.1/dist/module/lib/evaluation/handlers/inputDescriptorFilterEvaluationHandler.js
import Ajv from 'ajv';
import addFormats from 'ajv-formats';
import jp from 'jsonpath';
import { Status } from '../../ConstraintUtils';
import PexMessages from '../../types/Messages';
import { JsonPathUtils } from '../../utils';
import { AbstractEvaluationHandler } from './abstractEvaluationHandler';
export class InputDescriptorFilterEvaluationHandler extends AbstractEvaluationHandler {
    constructor(client) {
        super(client);
    }
    getName() {
        return 'FilterEvaluation';
    }
    handle(pd, wrappedVcs) {
        const fields = jp.nodes(pd, '$..fields[*]');
        wrappedVcs.forEach((wvc, vcIndex) => {
            this.createNoFieldResults(pd, vcIndex, wvc);
            fields.forEach((field) => {
                let inputField = [];
                if (field.value.path) {
                    inputField = JsonPathUtils.extractInputField(wvc.credential, field.value.path);
                }
                let resultFound = false;
                for (const inputFieldKey of inputField) {
                    if (this.evaluateFilter(inputFieldKey, field.value)) {
                        resultFound = true;
                        const payload = { result: { ...inputField[0] }, valid: true, format: wvc.format };
                        this.getResults().push({
                            ...this.createResultObject(jp.stringify(field.path.slice(0, 3)), vcIndex, payload),
                        });
                    }
                }
                if (!resultFound) {
                    if (!inputField.length) {
                        const payload = { valid: false, format: wvc.format };
                        this.createResponse(field, vcIndex, payload, PexMessages.INPUT_CANDIDATE_DOESNT_CONTAIN_PROPERTY);
                    }
                    else {
                        const payload = { result: { ...inputField[0] }, valid: false, format: wvc.format };
                        this.createResponse(field, vcIndex, payload, PexMessages.INPUT_CANDIDATE_FAILED_FILTER_EVALUATION);
                    }
                }
            });
        });
        this.updatePresentationSubmission(pd);
    }
    createNoFieldResults(pd, vcIndex, credential) {
        // PresentationDefinitionV2 is the common denominator
        const noFields = pd.input_descriptors
            .map((inDesc, index) => {
            return { index, inDesc };
        })
            .filter((el) => el.inDesc.constraints?.fields === undefined || el.inDesc.constraints?.fields?.length === 0);
        noFields.forEach((noField) => {
            const payload = { result: [], valid: true, format: credential.format };
            this.getResults().push({
                ...this.createResultObject(`$.input_descriptors[${noField.index}]`, vcIndex, payload),
            });
        });
    }
    createResponse(field, vcIndex, payload, message) {
        this.getResults().push({
            // TODO: why does this code assume a path to contain certain elements in the path?
            ...this.createResultObject(jp.stringify(field.path.slice(0, 3)), vcIndex, payload),
            ['status']: Status.ERROR,
            ['message']: message,
        });
    }
    createResultObject(path, vcIndex, payload) {
        return {
            input_descriptor_path: path,
            verifiable_credential_path: `$[${vcIndex}]`,
            evaluator: this.getName(),
            status: Status.INFO,
            message: PexMessages.INPUT_CANDIDATE_IS_ELIGIBLE_FOR_PRESENTATION_SUBMISSION,
            payload,
        };
    }
    evaluateFilter(result, field) {
        if (field.filter?.format && field.filter.format === 'date') {
            this.transformDateFormat(result);
        }
        const ajv = new Ajv({ verbose: true, allowUnionTypes: true, allErrors: true, strict: false });
        addFormats(ajv);
        if (field.filter) {
            return ajv.validate(field.filter, result.value);
        }
        return true;
    }
    transformDateFormat(result) {
        const date = new Date(result.value);
        let month = date.getUTCMonth() + 1 + '';
        if (month.length === 1) {
            month = '0' + month;
        }
        let day = date.getUTCDate() + '';
        if (day.length === 1) {
            day = '0' + day;
        }
        result.value = date.getUTCFullYear() + '-' + month + '-' + day;
        result.value = date.toISOString().substring(0, date.toISOString().indexOf('T'));
    }
}
