// @sphereon/pex@2.0.1/dist/module/lib/types/SSITypesBuilder.js
import { CredentialMapper, } from '@sphereon/ssi-types';
import { definitionVersionDiscovery, JsonPathUtils } from '../utils';
import { InternalPresentationDefinitionV1, InternalPresentationDefinitionV2, PEVersion, } from './Internal.types';
export class SSITypesBuilder {
    static modelEntityToInternalPresentationDefinitionV1(p) {
        const pd = SSITypesBuilder.createCopyAndModifyPresentationDefinition(p);
        return new InternalPresentationDefinitionV1(pd.id, pd.input_descriptors, pd.format, pd.name, pd.purpose, pd.submission_requirements);
    }
    static modelEntityInternalPresentationDefinitionV2(p) {
        const pd = SSITypesBuilder.createCopyAndModifyPresentationDefinition(p);
        return new InternalPresentationDefinitionV2(pd.id, pd.input_descriptors, pd.format, pd.frame, pd.name, pd.purpose, pd.submission_requirements);
    }
    static createCopyAndModifyPresentationDefinition(p) {
        const pd = JSON.parse(JSON.stringify(p));
        JsonPathUtils.changePropertyNameRecursively(pd, '_const', 'const');
        JsonPathUtils.changePropertyNameRecursively(pd, '_enum', 'enum');
        JsonPathUtils.changeSpecialPathsRecursively(pd);
        return pd;
    }
    static mapExternalVerifiablePresentationToWrappedVP(presentation) {
        return CredentialMapper.toWrappedVerifiablePresentation(presentation);
    }
    static mapExternalVerifiableCredentialsToWrappedVcs(verifiableCredentials) {
        return CredentialMapper.toWrappedVerifiableCredentials(Array.isArray(verifiableCredentials) ? verifiableCredentials : [verifiableCredentials]);
    }
    static toInternalPresentationDefinition(presentationDefinition) {
        const presentationDefinitionCopy = JSON.parse(JSON.stringify(presentationDefinition));
        const versionResult = definitionVersionDiscovery(presentationDefinitionCopy);
        if (versionResult.error)
            throw versionResult.error;
        if (versionResult.version == PEVersion.v1) {
            return SSITypesBuilder.modelEntityToInternalPresentationDefinitionV1(presentationDefinitionCopy);
        }
        return SSITypesBuilder.modelEntityInternalPresentationDefinitionV2(presentationDefinitionCopy);
    }
}
