// @sphereon/pex@2.0.1/dist/module/lib/validation/bundlers/frameVB.js
import { ObjectUtils } from '../../utils';
import { ValidationBundler } from './validationBundler';
export class FrameVB extends ValidationBundler {
    frameIsValidMsg = 'frame value is not valid';
    constructor(parentTag) {
        super(parentTag, 'frame');
    }
    getValidations(frame) {
        let validations = [];
        validations = [...validations, ...this.getMyValidations(frame)];
        return validations;
    }
    getMyValidations(frame) {
        return [
            {
                tag: this.getMyTag(),
                target: frame,
                predicate: FrameVB.frameIsValid,
                message: this.frameIsValidMsg,
            },
        ];
    }
    getMyTag() {
        return this.parentTag + '.' + this.myTag;
    }
    /**
     * this is based on https://github.com/digitalbazaar/jsonld.js/blob/main/lib/frame.js
     * @param frame
     */
    static frameIsValid(frame) {
        if (!frame || Array.isArray(frame) || !(typeof frame === 'object')) {
            return false;
        }
        const fr = frame;
        if (fr && fr['@id']) {
            for (const id of ObjectUtils.asArray(frame['@id'])) {
                // @id must be wildcard or an IRI
                if (!(ObjectUtils.isObject(id) || ObjectUtils.isUrlAbsolute(id)) || (ObjectUtils.isString(id) && id.indexOf('_:') === 0)) {
                    return false;
                }
            }
        }
        //
        if (fr['@types']) {
            for (const type of ObjectUtils.asArray(frame['@types'])) {
                // @id must be wildcard or an IRI
                if (!(ObjectUtils.isObject(type) || ObjectUtils.isUrlAbsolute(type)) || (ObjectUtils.isString(type) && type.indexOf('_:') === 0)) {
                    return false;
                }
            }
        }
        return true;
    }
}
