// @sphereon/pex@2.0.1/dist/module/lib/validation/bundlers/inputDescriptorsV1VB.js
import { ObjectValidationUtils } from '../../utils';
import { ConstraintsVB } from './constraintsVB';
import { ValidationBundler } from './validationBundler';
export class InputDescriptorsV1VB extends ValidationBundler {
    idMustBeNonEmptyStringMsg = 'input descriptor id must be non-empty string';
    nameShouldBeNonEmptyStringMsg = 'input descriptor name should be non-empty string';
    purposeShouldBeNonEmptyStringMsg = 'input descriptor purpose should be non-empty string';
    shouldHaveValidSchemaURIMsg = 'schema should have valid URI';
    constructor(parentTag) {
        super(parentTag, 'input_descriptor');
    }
    getValidations(inputDescriptors) {
        let validations = [];
        validations.push({
            tag: this.getTag(),
            target: inputDescriptors,
            predicate: (inDescs) => this.shouldHaveUniqueIds(inDescs),
            message: 'input descriptor ids must be unique',
        }, {
            tag: this.getTag(),
            target: inputDescriptors,
            predicate: (inDescs) => this.shouldHaveUniqueFieldsIds(inDescs),
            message: 'fields id must be unique',
        });
        inputDescriptors.forEach((inputDescriptor, inDescInd) => {
            validations = [
                ...validations,
                ...this.getValidationFor(inputDescriptor, inDescInd),
                ...this.constraintsValidations(inputDescriptor, inDescInd),
            ];
        });
        return validations;
    }
    getValidationFor(inputDescriptor, inDescInd) {
        return [
            {
                tag: this.getMyTag(inDescInd),
                target: inputDescriptor,
                predicate: (inDesc) => ObjectValidationUtils.nonEmptyString(inDesc?.id),
                message: this.idMustBeNonEmptyStringMsg,
            },
            {
                tag: this.getMyTag(inDescInd),
                target: inputDescriptor,
                predicate: this.isValidSchema(),
                message: this.shouldHaveValidSchemaURIMsg,
            },
            {
                tag: this.getMyTag(inDescInd),
                target: inputDescriptor,
                predicate: (inDesc) => ObjectValidationUtils.optionalNonEmptyString(inDesc?.name),
                message: this.nameShouldBeNonEmptyStringMsg,
            },
            {
                tag: this.getMyTag(inDescInd),
                target: inputDescriptor,
                predicate: (inDesc) => ObjectValidationUtils.optionalNonEmptyString(inDesc?.purpose),
                message: this.purposeShouldBeNonEmptyStringMsg,
            },
        ];
    }
    shouldHaveUniqueFieldsIds(inputDescriptors) {
        const nonUniqueInputDescriptorFieldsIds = [];
        const uniqueInputDescriptorFieldsIds = new Set();
        const tmp = [];
        inputDescriptors
            .map((e) => e.constraints?.fields)
            .forEach((e) => {
            if (e) {
                tmp.push(...e);
            }
        });
        tmp.forEach((e) => {
            if (e.id) {
                nonUniqueInputDescriptorFieldsIds.push(e.id);
            }
        });
        nonUniqueInputDescriptorFieldsIds.forEach((id) => uniqueInputDescriptorFieldsIds.add(id));
        return nonUniqueInputDescriptorFieldsIds.length === uniqueInputDescriptorFieldsIds.size;
    }
    shouldHaveUniqueIds(inputDescriptors) {
        const nonUniqueInputDescriptorIds = [];
        const uniqueInputDescriptorIds = new Set();
        inputDescriptors.forEach((e) => nonUniqueInputDescriptorIds.push(e.id));
        nonUniqueInputDescriptorIds.forEach((e) => uniqueInputDescriptorIds.add(e));
        return nonUniqueInputDescriptorIds.length === uniqueInputDescriptorIds.size;
    }
    getMyTag(srInd) {
        // TODO extract to make it generic
        return this.parentTag + '.' + this.myTag + '[' + srInd + ']';
    }
    isValidSchema() {
        // TODO extract to generic util or use built-in method
        return (inDesc) => {
            return (inDesc.schema.filter((schema) => this.isAValidURI(schema.uri) && (schema.required == null || typeof schema.required == 'boolean'))
                .length > 0);
        };
    }
    isAValidURI(uri) {
        if (!uri) {
            return false;
        }
        else if (!ObjectValidationUtils.nonEmptyString(uri)) {
            return false;
        }
        if (uri.startsWith('http://') || uri.startsWith('https://')) {
            try {
                new URL(uri);
            }
            catch (err) {
                return ObjectValidationUtils.isValidDIDURI(uri);
            }
        }
        return true;
    }
    constraintsValidations(inputDescriptor, inDescInd) {
        if (inputDescriptor.constraints) {
            return new ConstraintsVB(this.getMyTag(inDescInd)).getValidations(inputDescriptor.constraints);
        }
        return [];
    }
}
