// @sphereon/pex@2.0.1/dist/module/lib/validation/bundlers/submissionRequirementVB.js
import { ValidationBundler } from './validationBundler';
export class SubmissionRequirementVB extends ValidationBundler {
    ruleIsMandatoryMsg = 'rule is a mandatory field';
    needsOneFromOrFromNestedMsg = 'needs exactly one of from or from_nested';
    fromNestedShouldBeArrayMsg = 'The value of the from_nested property MUST be an array';
    isCountPositiveIntMsg = 'count must be a practical positive number';
    isMinPositiveIntMsg = 'min must be a practical positive number';
    isMaxPositiveIntMsg = 'max must be a practical positive number';
    ruleShouldBePickOrAllMsg = 'rule should be either pick or all';
    constructor(parentTag) {
        super(parentTag, 'submission_requirements');
    }
    getValidations(srs) {
        let validations = [];
        if (srs != null && srs.length > 0) {
            for (let srInd = 0; srInd < srs.length; srInd++) {
                validations = [...validations, ...this.getMyValidations(srInd, srs), ...this.getSubValidations(srInd, srs)];
            }
        }
        return validations;
    }
    getMyValidations(srInd, srs) {
        return [
            {
                tag: this.getMyTag(srInd),
                target: srs[srInd],
                predicate: SubmissionRequirementVB.ruleIsMandatory,
                message: this.ruleIsMandatoryMsg,
            },
            {
                tag: this.getMyTag(srInd),
                target: srs[srInd],
                predicate: SubmissionRequirementVB.needsOneFromOrFromNested,
                message: this.needsOneFromOrFromNestedMsg,
            },
            {
                tag: this.getMyTag(srInd),
                target: srs[srInd],
                predicate: SubmissionRequirementVB.fromNestedShouldBeArray,
                message: this.fromNestedShouldBeArrayMsg,
            },
            // TODO Validation 4.2.1.E All objects in from_nested should be of type SubmissionRequirement
            //      See if it can be implemented in pe-api yamls. currently in typescript type of this variable is 'any'
            //      i.e. from_nested?: Array<object>;
            {
                tag: this.getMyTag(srInd),
                target: srs[srInd],
                predicate: this.isCountPositiveInt,
                message: this.isCountPositiveIntMsg,
            },
            {
                tag: this.getMyTag(srInd),
                target: srs[srInd],
                predicate: this.isMinPositiveInt,
                message: this.isMinPositiveIntMsg,
            },
            {
                tag: this.getMyTag(srInd),
                target: srs[srInd],
                predicate: this.isMaxPositiveInt,
                message: this.isMaxPositiveIntMsg,
            },
            {
                tag: this.getMyTag(srInd),
                target: srs[srInd],
                predicate: SubmissionRequirementVB.ruleShouldBePickOrAll,
                message: this.ruleShouldBePickOrAllMsg,
            }, // Validation 4.2.4
        ];
    }
    getMyTag(srInd) {
        // TODO extract to make it generic
        return this.parentTag + '.' + this.myTag + '[' + srInd + ']';
    }
    getSubValidations(srInd, srs) {
        const fromNested = srs[srInd].from_nested;
        return fromNested != null ? new SubmissionRequirementVB(this.getFromNestedTag(srInd)).getValidations(fromNested) : [];
    }
    getFromNestedTag(srInd) {
        return this.getMyTag(srInd) + '.' + 'from_nested';
    }
    isCountPositiveInt(sr) {
        return sr.rule !== 'pick' || sr.count == null || 0 < sr.count;
    }
    isMinPositiveInt(sr) {
        return sr.rule !== 'pick' || sr.min == null || 0 <= sr.min;
    }
    isMaxPositiveInt(sr) {
        return sr.rule !== 'pick' || sr.max == null || 0 < sr.max;
    }
    static ruleIsMandatory(sr) {
        return sr.rule != null;
    }
    static needsOneFromOrFromNested(sr) {
        return (sr.from == null) !== (sr.from_nested == null); // XOR
    }
    static fromNestedShouldBeArray(sr) {
        return sr.from_nested == null || Array.isArray(sr.from_nested);
    }
    static ruleShouldBePickOrAll(sr) {
        return sr.rule === 'pick' || sr.rule === 'all';
    }
}
